typescript/infra/src/funding/adapters/SealevelFundingAdapter.ts

```diff
@@ -0,0 +1,304 @@
+import {
+  ComputeBudgetProgram,
+  LAMPORTS_PER_SOL,
+  PublicKey,
+  SystemProgram,
+  Transaction,
+  sendAndConfirmTransaction,
+} from '@solana/web3.js';
+import { Logger } from 'pino';
+import { Gauge } from 'prom-client';
+
+import {
+  BaseSealevelAdapter,
+  ChainMap,
+  ChainName,
+  MultiProtocolProvider,
+  SEALEVEL_PRIORITY_FEES,
+  SealevelIgpAdapter,
+} from '@hyperlane-xyz/sdk';
+
+import { Contexts } from '../../../config/contexts.js';
+import { BaseAgentKey } from '../../agents/keys.js';
+import { DeployEnvironment } from '../../config/environment.js';
+import { FundableRole } from '../../roles.js';
+import { FunderAddresses } from '../types.js';
+
+import { IFundingAdapter } from './IFundingAdapter.js';
+
+/**
+ * Sealevel-specific implementation of the funding adapter
+ * Extends BaseSealevelAdapter to leverage Sealevel-specific functionality
+ */
+export class SealevelFundingAdapter
+  extends BaseSealevelAdapter
+  implements IFundingAdapter
+{
+  private igpAdapter: SealevelIgpAdapter;
+  private igpProgramIds: ChainMap<string>;
+
+  constructor(
+    chainName: ChainName,
+    multiProvider: MultiProtocolProvider,
+    environment: DeployEnvironment,
+    private readonly context: Contexts,
+    private readonly fundingAddresses: FunderAddresses,
+    public readonly logger: Logger,
+    private readonly walletBalanceGauge: Gauge<string>,
+  ) {
+    super(chainName, multiProvider, fundingAddresses, logger);
+    this.logger = logger;
+    // TODO: this should be a config that is passed in
+    this.igpProgramIds = {
+      solanamainnet: 'BhNcatUDC2D5JTyeaqrdSukiVFsEHK7e3hVmKMztwefv',
+      sonicsvm: '3kd3K1yymsfYoqzNnktLwpKXKeB9Y2PRf9rfb31k4o13',
+      soon: '5AAMBemjUrAprKJzi22Si9pyYhK8vTr7Jpti7LEXo6d8',
+      eclipsemainnet: 'Hs7KVBU67nBnWhDPZkEFwWqrFMUfJbmY2DQ4gmCZfaZp',
+    };
+    this.igpAdapter = new SealevelIgpAdapter(
+      this.chainName,
+      this.multiProvider,
+      {
+        igp: this.fundingAddresses.interchainGasPaymaster,
+        programId: this.getIgpProgramId(),
+      },
+    );
+  }
+
+  async getBalance(address: string): Promise<bigint> {
+    const response = await this.getProvider().getBalanceAndContext(
+      new PublicKey(address),
+    );
+    // Get the raw value from the RPC response before it's converted to number
+    return BigInt(response.value);
+  }
+
+  async getFundingAmount(
+    address: string,
+    desiredBalance: number,
```

Comment:
Maybe the desired balances by chain can be provided as a config object in the constructor? wdyt?